### ISSUE TITLE
Reasons for the huge difference in performance between Mesh_boolean and CSGTree

### ISSUE THREAD

**Issue by gdwei** [libigl/libigl#2377]
### Discussed in https://github.com/libigl/libigl/discussions/2376

<div type='discussions-op-text'>

<sup>Originally posted by **gdwei** April 19, 2024</sup>
I am using the Mesh_boolean and CSGTree to do the same thing, that is, 

```
Eigen::MatrixXd Vout_a;
Eigen::MatrixXi Fout_a;
Eigen::VectorXi J_a;
igl::copyleft::cgal::mesh_boolean(Va, Fa, Vb, Fb, igl::MESH_BOOLEAN_TYPE_INTERSECT, Vout_a, Fout_a, J_a);
        
Eigen::MatrixXd VertexOut;
Eigen::MatrixXi FaceOut;
Eigen::VectorXi J;
igl::copyleft::cgal::mesh_boolean(Vc, Fc, Vout_a, Fout_a, igl::MeshBooleanType::MESH_BOOLEAN_TYPE_MINUS, VertexOut, FaceOut, J);

```

and using CSGTree like 

```
const igl::copyleft::cgal::CSGTree csgtree{ 
    {Vc, Fc}, 
    {{Va, Fa}, {Vb, Fb}, igl::MESH_BOOLEAN_TYPE_INTERSECT}, 
    igl::MeshBooleanType::MESH_BOOLEAN_TYPE_MINUS
 };
Eigen::MatrixXd VertexOut = csgtree.cast_V<Eigen::MatrixXd>();
Eigen::MatrixXi FaceOut = csgtree.F();
```

I understand that using Mesh_boolean several times would cast exact numbers to float point numbers and causing some differences. But the time consuming by CSGTree is almost twice of mesh_boolean. 

Is it normal? And what are the reasons for such big differences?</div>

**Comment by alecjacobson** [comment]
Yeah, I think it's reasonably expected. CSGTree keeps things exact the whole time, but you're call to mesh_boolean twice is rounding to float in between.

**Comment by gdwei** [comment]
> Yeah, I think it's reasonably expected. CSGTree keeps things exact the whole time, but you're call to mesh_boolean twice is rounding to float in between.

But the rounding operations are expected to take extra time compared to CSGTree that keep exact all the time,  isn't it?

**Comment by alecjacobson** [comment]
The rounding operation itself is pretty cheap. Rounding keeps the output of each boolean in floats, so the exact types don't grow in size.

Without rounding, you can think of each boolean operation as "doubling" the size and cost of doing arithmetic on each mesh vertex position.

**Comment by gdwei** [comment]
> The rounding operation itself is pretty cheap. Rounding keeps the output of each boolean in floats, so the exact types don't grow in size.
> 
> Without rounding, you can think of each boolean operation as "doubling" the size and cost of doing arithmetic on each mesh vertex position.

Thanks, it makes sense to me now. But still, what is the best\fastest way to do the boolean operations above? I found that there exist such a function

```
IGL_INLINE bool mesh_boolean(
          const Eigen::MatrixBase<DerivedVV > & VV,
          const Eigen::MatrixBase<DerivedFF > & FF,
          const Eigen::MatrixBase<Derivedsizes> & sizes,
          const std::function<int(const Eigen::Matrix<int,1,Eigen::Dynamic>) >& wind_num_op,
          const std::function<int(const int, const int)> & keep,
          Eigen::PlainObjectBase<DerivedVC > & VC,
          Eigen::PlainObjectBase<DerivedFC > & FC,
          Eigen::PlainObjectBase<DerivedJ > & J);
```

which seems will be faster when involve multiple boolean operations among several meshes. 
However, I can't figure out how to use it. It will be much appreciated if you can show me the way, thanks!
